### ISSUE TITLE
The component styled.div with the id of "sc-fzqAui" has been created dynamically. You may see this warning because you've called styled inside another component. To resolve this only create new StyledComponents outside of any render method and function component.

### ISSUE THREAD

**Issue by eugle** [styled-components/styled-components#3117]
The component styled.div with the id of "sc-fzqAui" has been created dynamically.
You may see this warning because you've called styled inside another component.
To resolve this only create new StyledComponents outside of any render method and function component.

![image](https://user-images.githubusercontent.com/24447931/79686907-4abeee00-8276-11ea-9931-1793236ea99b.png)

Is there any way to get rid of this warning? There will be a warning in the development environment, and everything is normal in the production environment

This is my actual component code

![image](https://user-images.githubusercontent.com/24447931/79686957-a38e8680-8276-11ea-9c00-80155ddae923.png)

Based on the component's properties, I wrapped an NPM package,

[@eugle/oneui](https://www.npmjs.com/package/@eugle/oneui)

**Comment by CyrilQuandalle** [comment]
Hi,
Can you show us where you've declared your components ? I've generaly got this warning when my styled components were declared inside a function component.

**Comment by alac1984** [comment]
> Hi,
> Can you show us where you've declared your components ? I've generaly got this warning when my styled components were declared inside a function component.

I'm getting the same. So, what is the right way to do it? Where should I declare styled components?

**Comment by CyrilQuandalle** [comment]
Hi,
Like that :
```
import styled from 'styled-components'

const StyledSpan = styled.span`
    color: blue;
`
const MyExampleComponent = () =>{
    return <StyledSpan>Test</StyledSpan>
}
```

I think you're doing it like that:
```
import styled from 'styled-components'

const MyExampleComponent = () =>{

    const StyledSpan = styled.span`
        color: blue;
    `
    return <StyledSpan>Test</StyledSpan>
}
```

That's why you've got the warning.

**Comment by eugle** [comment]
`function Header() {
	const { DefaultColor } = useContext(GlobalContext);
	
	const Nav = styled.div`
		div > a.active > div {
			color: ${DefaultColor?.success};
		}
	`;
	return (`

**Comment by eugle** [comment]
`import { GlobalContext } from '/imports/ui/layout/App';
import styled from 'styled-components';

function Header() {
	const { DefaultColor } = useContext(GlobalContext);

	const Nav = styled.div `
		div > a.active > div {
			color: ${DefaultColor?.success};
		}
	`;

`

**Comment by eugle** [comment]
@CyrilQuandalle  Thank you very much for pointing out the problem, and I did, my code

![image](https://user-images.githubusercontent.com/24447931/80056666-4edd5b00-8557-11ea-9824-5b59ee1cdeff.png)

Is there a good way to get around this because I want to get the parent context parameter in the component

Once again thank you

**Comment by CyrilQuandalle** [comment]
You can do that like that :
```
const Nav = styled.div`
    div > a.active > div {
        color: ${props => props.color}
     }
`
function Header() {
    const {DefaultColor} = useContext(GlobalContext)
    return <Nav color={DefaultColor?.success} />
}
```

**Comment by eugle** [comment]
wow, So simple, why didn't I think, thank you for your kind help, I will try to get rid of this mistake

**Comment by CyrilQuandalle** [comment]
https://styled-components.com/docs/basics#passed-props

**Comment by eugle** [comment]
@CyrilQuandalle It's very kind of you to solve the problem once and for all

**Comment by praveendass** [comment]
@CyrilQuandalle Even I am facing the same issue. I like to keep my style objects in a separate JS file. 
For styles without any dynamic attributes, I don't face any issue.

But in the other case, I write something like the below:
export const getColumn = (ratio) => {
  return styled.div`  
      flex: ${ratio};  
  `
}

and use it like below:
                const Column = getColumn(columnRatio[index]);
                return(<Column key={index} >{header}</Column>) 

In this case, I face the issue under discussion. What would be the workaround?

**Comment by CyrilQuandalle** [comment]
Hi @praveendass ,

I never used it like that ;-). Maybe move `const Column = getColumn(columnRatio[index]);` outside the function component. Just above the component in the same file for example.

**Comment by praveendass** [comment]
@CyrilQuandalle Hey thanks for the quick response!
 That wouldn't do, because the value of ratio keeps changing as I iterate.

const getColumn = (ratio) => {
    return (styled.div`
        flex: ${ratio}; 
    `)
} 
const TableHeaderComponent = ({headers, columnRatio}) => {

    return ( 
        <TableHeader>
            {
            headers.map((header, index) => {
                const Column = getColumn(columnRatio[index]);
                return(<Column key={index} >{header}</Column>)
            }) }

        </TableHeader>
     );
}

Any suggestion on how I can get around this?

**Comment by CyrilQuandalle** [comment]
I dont know then. Why do you need a function to render your  divs as you can just pass a props for the ratio to it ?
like that : `const Column = styled.div`flex: ${props => props.ratio}``
`const TableHeaderComponent = ({headers, columnRatio}) => {
return ( 
    <TableHeader>
        {
        headers.map((header, index) => {
            return(<Column key={index} ratio={columnRatio} >{header}</Column>)
        }) }

    </TableHeader>
 );`

**Comment by praveendass** [comment]
@CyrilQuandalle Just figured it out... :D
I wasn't aware that styled-components are HOC wrapped over component...
I was in the idea that it was a JS object and didn't have props attribute for itself
Well, now I know! 
https://stackoverflow.com/questions/61874381/the-component-styled-has-been-created-dynamically-you-may-see-this-warning-be

**Comment by heymartinadams** [comment]
Sorry for adding another comment here — I’m not understanding what to do. Receiving the same warning, yet nothing I do can fix this. Like @eugle, I, too, want to keep my styled-components in *separate JS files*, since many SCs are used as building blocks over and over again.

How can I refactor the following?

## Section.js
```js
export default styled.div`
	position: relative;
	width: 100%;
`
```

## Homepage.js
```js
import Section from '@/Section'

const Homepage = () => (
	<Section>
		Hello, world!
	</Section>
)

export default Homepage
```

Tried various changes to no avail.

**Comment by CyrilQuandalle** [comment]
Hi @heymartinadams ,

What warning do you have ?
Thats strang though, I replecated your code in the following codesandbox and I have no warning : 
[https://codesandbox.io/s/styled-components-vi0g5?file=/index.js](url)

**Comment by heymartinadams** [comment]
Here’s the error message, @CyrilQuandalle:
```
The component Section with the id of "Section-sc-1fu27j3-0" has been created dynamically.
You may see this warning because you've called styled inside another component.
To resolve this only create new StyledComponents outside of any render method and function component.
```
(for each and every SC)

Mmh... my situation is probably made more complex in that I’m using Preact (instead of React) as well as the NextJS framework. Could that perhaps be the cause of it?

For example, in the `_document.js` file (which gets pre-loaded before the actual homepage gets called):

```js
import Document from 'next/document'
import { ServerStyleSheet } from 'styled-components'

export default class MyDocument extends Document {
	// Not sure if needed
	static async getInitialProps(ctx) {
		const sheet = new ServerStyleSheet()
		const originalRenderPage = ctx.renderPage

		try {
			ctx.renderPage = () =>
				originalRenderPage({
					enhanceApp: App => props => sheet.collectStyles(<App {...props} />)
				})

			const initialProps = await Document.getInitialProps(ctx)
			return {
				...initialProps,
				styles: (
					<>
						{initialProps.styles}
						{sheet.getStyleElement()}
					</>
				)
			}
		} finally {
			sheet.seal()
		}
	}
}
```

**Comment by CyrilQuandalle** [comment]
Yes it could caused by Preact or NextJS. I just found that I dont know if it fits your case :

[https://github.com/styled-components/styled-components/issues/3174](url)

**Comment by heymartinadams** [comment]
Thank you, @CyrilQuandalle! Subscribed to the corresponding [Preact issue](https://github.com/preactjs/preact/issues/2574).

At least it seems it’s only a warning trigger that’s at issue, not an actual problem with the code or execution.

**Comment by felipemotabr** [comment]
> Hi,
> Like that :
> 
> ```
> import styled from 'styled-components'
> 
> const StyledSpan = styled.span`
>     color: blue;
> `
> const MyExampleComponent = () =>{
>     return <StyledSpan>Test</StyledSpan>
> }
> ```
> 
> I think you're doing it like that:
> 
> ```
> import styled from 'styled-components'
> 
> const MyExampleComponent = () =>{
> 
>     const StyledSpan = styled.span`
>         color: blue;
>     `
>     return <StyledSpan>Test</StyledSpan>
> }
> ```
> 
> That's why you've got the warning.

This is work. Thank you

**Comment by hydego17** [comment]
Hi, Im getting the same warning message when I tried to user some components from [react styled loaders](https://github.com/saisandeepvaddi/styled-loaders-react)

`import React from "react";
import { Bar } from "styled-loaders-react";

const Loader = () => {
  return <Bar duration="1.2s" />;
};

export default Loader;`

is there any solution?

**Comment by sarahrara18** [comment]
Hello, i have the same issue while using react-data-table-component ,i posted my problem here  jbetancur/react-data-table-component#840 ,  the problem is i didn't use styledComponent at all

**Comment by theMyth721** [comment]
The only 

> wow, So simple, why didn't I think, thank you for your kind help, I will try to get rid of this mistake

When I tried to do something similar, my whole browser crashed. Im facing the same problem, Firefox gets uncaught out of memory exception

**Comment by DungVietLe** [comment]
excusme~  . some one help me 
![image](https://user-images.githubusercontent.com/74787642/135989634-eb243431-1406-4be7-a7fa-9a1f50e85e06.png)

in my function component there is no styled but why does it say warning ?

![image](https://user-images.githubusercontent.com/74787642/135989804-0c50a7cf-fb56-4680-8022-35305e2ce4c7.png)

**Comment by yeasinjabed2** [comment]
**My problem with SVG , Any solutions ?**
`export const Icon = ({ svg, width, height, fill }) => {
  const Svg = styled(svg)`
    width: ${width};
    height: ${height};

    & path {
      fill: ${fill};
    }
  `;

  return <Svg />;
};`

**Comment by yeasinjabed2** [comment]
**My problem with SVG , any solutions ?**
![image](https://user-images.githubusercontent.com/30934078/152019037-95547d66-657b-4ada-be67-d182d0054231.png)

**Comment by yeasinjabed2** [comment]
> Hi, Can you show us where you've declared your components ? I've generaly got this warning when my styled components were declared inside a function component.

**My problem with SVG, any solutions ?**
![image](https://user-images.githubusercontent.com/30934078/152019327-8e1b2652-32ea-4e58-916c-9d526a966638.png)

**Comment by pkaufi** [comment]
You can try with

```
import { cloneElement, HTMLProps } from "react";
.
.
.
export const Icon = styled(({
                         name,
                         color,
                         opacity,
                         size = "default",
                         inverted,
                        ...props }: IconProps) => cloneElement(iconOptions[name], props)
)`
    fill: ${({ color }) => color};
    opacity: ${({ opacity }) => opacity};

    ${({ inverted }) => inverted &&
    `
    -webkit-transform: scaleY(-1);
    transform: scaleY(-1);
  `}

    ${({ size }) => size === "md" &&
    `
    width: 9px;
  `}

  ${({ size }) => size === "sm" &&
    `
    width: 6.75px;
  `}
`;
```

**Comment by skmainwal** [comment]
> Hi, Like that :
> 
> ```
> import styled from 'styled-components'
> 
> const StyledSpan = styled.span`
>     color: blue;
> `
> const MyExampleComponent = () =>{
>     return <StyledSpan>Test</StyledSpan>
> }
> ```
> 
> I think you're doing it like that:
> 
> ```
> import styled from 'styled-components'
> 
> const MyExampleComponent = () =>{
> 
>     const StyledSpan = styled.span`
>         color: blue;
>     `
>     return <StyledSpan>Test</StyledSpan>
> }
> ```
> 
> That's why you've got the warning.

Hi ,I have written the code like this but I am getting an error

**Comment by jvnlwn** [comment]
Came across the same error and it did not appear that the styled component was actually being created dynamically, at least not to my un(styled)trained eyes.

[Here's a repro sandbox](https://codesandbox.io/s/styled-components-base-forked-df586m) on version 5.3.5 of styled-components in case this helps. The warning was not thrown on version 5.0.0-beta which is currently the default version used in the [base codesandbox provided](https://codesandbox.io/s/styled-components-base-gc367) when creating a new issue.

The issue comes from requiring a styled component within the body of the functional component, like so:
```javascript
const Consumer = ({ componentName }) => {
  const Component = require(`./components`)[componentName];
  return <Component />;
};

```

A workaround is to import all components outside of the functional component:
```javascript
import * as Components from "./components"

const Consumer = ({ componentName }) => {
  const Component = Components[componentName]
  return <Component />;
};

```

Hope this helps someone.

**Comment by valdeirsilva12** [comment]
I solved it through this nextjs documentation in the link below

nextJs: [https://nextjs.org/docs/advanced-features/custom-document](url)

video: [https://youtu.be/lKD5Z0ORA54?t=1254](url)

**Comment by andideve** [comment]
how to ignore this problem?

**Comment by Henery002** [comment]
> You can do that like that :
> 
> ```
> const Nav = styled.div`
>     div > a.active > div {
>         color: ${props => props.color}
>      }
> `
> function Header() {
>     const {DefaultColor} = useContext(GlobalContext)
>     return <Nav color={DefaultColor?.success} />
> }
> ```

but where is your "props" from?
